spec/lib/hubspot/company_properties_spec.rb

```diff
@@ -14,158 +14,325 @@
-      HTTParty.get('https://api.hubapi.com/companies/v2/properties?hapikey=demo').parsed_response
-    end
-  end
+  describe "Properties" do
+    describe ".all" do
+      it "should return all properties" do
+        VCR.use_cassette "company_properties/all_properties" do
+          Hubspot::CompanyProperties.create!({
+            name: "fax_number",
+            label: "Fax Number",
+            description: "The company fax number.",
+            groupName: "companyinformation",
+            type: "string",
+            fieldType: "text"
+          })
 
-  before { Hubspot.configure(hapikey: 'demo') }
+          response = Hubspot::CompanyProperties.all
 
-  describe 'Properties' do
-    describe '.all' do
-      context 'with no filter' do
-        cassette 'company_properties/all_properties'
+          assert_hubspot_api_request(:get, "/companies/v2/properties")
 
-        it 'should return all properties' do
-          expect(Hubspot::CompanyProperties.all).to eql(example_properties)
+          company_property_names = response.map { |property| property["name"] }
+          expect(company_property_names).to include("fax_number")
+
+          Hubspot::CompanyProperties.delete!("fax_number")
         end
       end
 
-      let(:groups) { %w(calltrackinginfo emailinformation) }
-
-      context 'with included groups' do
-        cassette 'company_properties/properties_in_groups'
-
-        it 'should return properties for the specified group[s]' do
-          response = Hubspot::CompanyProperties.all({}, { include: groups })
-          response.each { |p| expect(groups.include?(p['groupName'])).to be true }
+      context "with included groups" do
+        it "should return properties for the specified group[s]" do
+          VCR.use_cassette("company_properties/filter_by_group") do
+            group_a = "socialmediainformation"
+            group_b = "companyinformation"
+
+            Hubspot::CompanyProperties.create!({
+              name: "instagram_handle",
+              label: "Instagram Handle",
+              description: "The company's Instagram handle.",
+              groupName: group_a,
+              type: "string",
+              fieldType: "text"
+            })
+            Hubspot::CompanyProperties.create!({
+              name: "fax_number",
+              label: "Fax Number",
+              description: "The company fax number.",
+              groupName: group_b,
+              type: "string",
+              fieldType: "text"
+            })
+
+            response = Hubspot::CompanyProperties.all({}, { include: [group_a] })
+
+            assert_hubspot_api_request(:get, "/companies/v2/properties")
+
+            group_names = response.map { |property| property["groupName"] }
+            expect(group_names).to include(group_a)
+            expect(group_names).not_to include(group_b)
+
+            Hubspot::CompanyProperties.delete!("instagram_handle")
```

Comment:
If the `assert` or any of the `expect`'s above fail do these get cleaned up? If not, does that break the test next iteration when the properties are created?

These may be better in a `before` and `after` block.

Reply:
> If the assert or any of the expect's above fail do these get cleaned up?

They do not. The test will exit before running these cleanup actions. That shouldn't be an issue most of the time except when the VCR recording is being updated. I'm still finding the best way to work with VCR and currently, my flow is to write the API calls, capture all the successful calls in a recording, and then write expectations against the recording.

 > If not, does that break the test next iteration when the properties are created?

Yes 😞 I have run into this where I create a record, the update or next step fails because the API behaves differently than I expected, and I then need to delete the created record before I can try to record the API interactions again. I don't know of a great way around this as we want each test to be self-sufficient and that means sometimes having to remove records manually. I use Postman when I need to remove a record.

> These may be better in a before and after block.

I'll play around with this and report back.

Reply:
I attempted to implement an `around` hook to handle the test setup and cleanup. Something like:
```
RSpec.describe CompanyProperties do
  describe ".all" do
    context "context" do
      around(:example) do |example|
        ... create data in HubSpot ...
        example.run
        .... delete data in HubSpot ...
      end

      it "runs" do
        ... assertion ...
      end
    end 
  end
end
```

This worked well for 99% of the tests in the CompanyProperties spec file but failed intermittently for one test. When run after the `.all` tests, the `delete!` test failed with the error message that VCR expected one HTTP request to the delete endpoint but found two calls being made. I'm fairly certain this had to do with the use of the `around` blocks but I couldn't identify the exact reason. I'm not sure if it's an RSpec issue or a VCR issue.

Since this approach hasn't worked (so far), an alternative is to do something like:
```
  ... destroy record that's about to be created and rescue if the record doesn't exist ...
  ... create record ...
```

This changes the test cadence from `test setup, run test, remove records created in test setup` to `delete any data that will conflict with test setup, test setup, run test`.

For now, I'm going to merge and keep moving as we continue to find the best way to test with VCR.